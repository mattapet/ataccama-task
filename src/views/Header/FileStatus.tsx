//#region imports
import React from 'react';
import styled from 'styled-components';
import { Button } from 'antd';
//#endregion

//#region Component interfaces
export interface Props {
  /**
   * Currently selected file. If no file is selected, the value is supposed
   * to be `null`.
   */
  readonly file: File | null;
  /**
   * Callback to clear the currently selected file.
   */
  readonly onClear: () => any;
}
//#endregion

//#region Styled
const Wrapper = styled.div`
  color: whitesmoke;
  * {
    color: whitesmoke;
    margin-right: 0.25em;
    margin-left: 0.25em;
  }
`;
//#endregion

/**
 * Displays currently selected file with an option to clear the selection.
 */
const FileStatus: React.FunctionComponent<Props> = props => (
  <Wrapper>
    <span>{props.file ? props.file.name : <em>No File Selected</em>}</span>
    <Button type="ghost" shape="round" onClick={props.onClear}>
      Clear
    </Button>
  </Wrapper>
);

export default FileStatus;
