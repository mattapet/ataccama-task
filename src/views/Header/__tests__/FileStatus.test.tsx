//#region imports
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import FileStatus from '../FileStatus';
//#endregion

let container: HTMLDivElement | null = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  container && document.body.removeChild(container);
  container = null;
});

it('renders correctly', () => {
  // Arrange
  // Act
  const component1 = renderer.create(
    <FileStatus file={null} onClear={() => {}} />
  );
  const component2 = renderer.create(
    <FileStatus file={new File([], 'test.json')} onClear={() => {}} />
  );

  // Assert
  expect(component1).toMatchSnapshot();
  expect(component2).toMatchSnapshot();
});

it('should display "No File Selected" when passed `null` file', () => {
  // Arrange

  // Act
  act(() => {
    ReactDOM.render(<FileStatus file={null} onClear={() => {}} />, container);
  });

  // Assert
  const em = container && container.querySelector('em');
  expect(em && em.textContent).toEqual('No File Selected');
});

it('should display name of the passed non-`null` file', async () => {
  // Arrage
  const fileName = 'TEST_FILE_NAME.json';

  // Act
  act(() => {
    ReactDOM.render(
      <FileStatus file={new File([], fileName)} onClear={() => {}} />,
      container
    );
  });

  // Assert
  const span = container && container.querySelector('span');
  expect(span && span.textContent).toEqual(fileName);
});

it('should invoke clear callback on click', async () => {
  // Arrange
  const fileName = 'TEST_FILE_NAME.json';
  const onClear = () => (cleared = true);
  let cleared = false;
  act(() => {
    ReactDOM.render(
      <FileStatus file={new File([], fileName)} onClear={onClear} />,
      container
    );
  });

  // Act
  const button = container && container.querySelector('button');
  act(() => {
    button && button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });

  // Assert
  expect(cleared).toBeTruthy();
});
