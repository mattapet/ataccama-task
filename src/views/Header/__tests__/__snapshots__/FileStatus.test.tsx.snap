// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bdVaJa NIJn"
>
  <span>
    <em>
      No File Selected
    </em>
  </span>
  <button
    className="ant-btn ant-btn-ghost ant-btn-round"
    onClick={[Function]}
    type="button"
  >
    <span>
      Clear
    </span>
  </button>
</div>
`;

exports[`renders correctly 2`] = `
<div
  className="sc-bdVaJa NIJn"
>
  <span>
    test.json
  </span>
  <button
    className="ant-btn ant-btn-ghost ant-btn-round"
    onClick={[Function]}
    type="button"
  >
    <span>
      Clear
    </span>
  </button>
</div>
`;
